Timeout API rework

andyross opened the pull request:
Big (but less so than the first version) API rework to make timeouts an opaque type, allowing for  their specification in alternate representations like 64 bit types, higher precision units and absolute/uptime values.

See #21305 for the RFC detailing the changes.

Commit:
timeout: Fix up API usage

Kernel timeouts have always been a 32 bit integer despite the
existence of generation macros, and existing code has been
inconsistent about using them.  Upcoming commits are going to make the
timeout arguments opaque, so fix things up to be rigorously correct.
Changes include:

+ Adding a K_TIMEOUT_EQ() macro for code that needs to compare timeout
  values for equality (e.g. with K_FOREVER or K_NO_WAIT).

+ Adding a k_msleep() synonym for k_sleep() which can continue to take
  integral arguments as k_sleep() moves away to timeout arguments.

+ Pervasively using the K_MSEC(), K_SECONDS(), et. al. macros to
  generate timeout arguments.

+ Removing the usage of K_NO_WAIT as the final argument to
  K_THREAD_DEFINE().  This is just a count of milliseconds and we need
  to use a zero.

This patch include no logic changes and should not affect generated
code at all.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Make timeout arguments an opaque type

Add a k_timeout_t type, and use it everywhere that kernel API
functions were accepting a millisecond timeout argument.  Instead of
forcing milliseconds everywhere (which are often not integrally
representable as system ticks), do the conversion to ticks at the
point where the timeout is created.  This avoids an extra unit
conversion in some application code, and allows us to express the
timeout in units other than milliseconds to achieve greater precision.

The existing K_MSEC() et. al. macros now return initializers for a
k_timeout_t.

The K_NO_WAIT and K_FOREVER constants have now become k_timeout_t
values, which means they cannot be operated on as integers.
Applications which have their own APIs that need to inspect these
vs. user-provided timeouts can now use a K_TIMEOUT_EQ() predicate to
test for equality.

Timer drivers, which receive an integer tick count in ther
z_clock_set_timeout() functions, now use the integer-valued
K_TICKS_FOREVER constant instead of K_FOREVER.

For the initial release, to preserve source compatibility, a
CONFIG_LEGACY_TIMEOUT_API kconfig is provided.  When true, the
k_timeout_t will remain a compatible 32 bit value that will work with
any legacy Zephyr application.

Some subsystems present timeout (or timeout-like) values to their own
users as APIs that would re-use the kernel's own constants and
conventions.  These will require some minor design work to adapt to
the new scheme (in most cases just using k_timeout_t directly in their
own API), and they have not been changed in this patch, instead
selecting CONFIG_LEGACY_TIMEOUT_API via kconfig.  These subsystems
include: CAN Bus, the Microbit display driver, I2S, LoRa modem
drivers, the UART Async API, Video hardware drivers, the console
subsystem, and the network buffer abstraction.

k_sleep() now takes a k_timeout_t argument, with a k_msleep() variant
provided that works identically to the original API.

Most of the changes here are just type/configuration management and
documentation, but there are logic changes in mempool, where a loop
that used a timeout numerically has been reworked using a new
z_timeout_end_calc() predicate.  Also in queue.c, a (when POLL was
enabled) a similar loop was needlessly used to try to retry the
k_poll() call after a spurious failure.  But k_poll() does not fail
spuriously, so the loop was removed.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Enable 64 bit timeout precision

Add a CONFIG_TIMEOUT_64BIT kconfig that, when selected, makes the
k_ticks_t used in timeout computations pervasively 64 bit.  This will
allow much longer timeouts and much faster (i.e. more precise) tick
rates.  It also enables the use of absolute (not delta) timeouts in an
upcoming commit.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Add absolute timeout APIs

Add support for "absolute" timeouts, which are expressed relative to
system uptime instead of deltas from current time.  These allow for
more race-resistant code to be written by allowing application code to
do a single timeout computation, once, and then reuse the timeout
value even if the thread wakes up and needs to suspend again later.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel.h: Add timeout APIs for higher precision units

Add timeout generator macros to the public API for micro/nanosecond
values and cycles.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/timeout: Add timeout remaining/expires APIs

Add tick-based (i.e. precision resistant) inspection APIs for kernel
timeouts visible via k_timer, k_delayed work and thread timeouts
(i.e. pended/sleeping threads).  These are each available in
"remaining" and "expires" variants returning time values relative to
current time and system start.  All have system calls where applicable
(i.e. everywhere but k_delayed_work, which is not a userspace API)

The pre-existing millisecond "remaining_get()" predicates for timer
and delayed work remain, but are expressed in terms of the newer
calls.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

dcpleung commented on include/sys_clock.h:
```diff
@@ -28,6 +28,59 @@
+
+typedef u32_t k_ticks_t;
+
+#define K_TICKS_FOREVER ((k_ticks_t) -1)
```
Since `k_ticks_t` is unsigned, `K_TICKS_FOREVER` is now of max integer value, instead of actual -1 (in old behavior). Would this affect the comparisons with `K_FOREVER`/`Z_FOREVER` when legacy timeout API is enabled?

andyross replied:
Not if everything is using the macros correctly, which is a compile error when legacy is not enabled.  The decision to make the 32 bit type unsigned and leave it that way in legacy mode was a compromise from the first version of this PR that I didn't change here.  It could be revisited I guess, but I fear this sort of thing can become a bikeshed.

Commit:
kernel/timeout: Add k_uptime_ticks() API

Add a call to get the system tick count as an official API (and
redefine the existing millisecond API in terms of it).  Sophisticated
applications need to be able to count ticks directly, and the newer
timeout API supports that.  Uptime should too, for symmetry.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andrewboie commented on drivers/led/lp5562.c:
```diff
@@ -68,12 +68,12 @@ LOG_MODULE_REGISTER(lp5562);
  * therefore (16 * 63) = 1008ms. We round it down to 1000ms to be on the safe
  * side.
  */
-#define LP5562_MAX_BLINK_PERIOD K_MSEC(1000)
```
Huh, I was surprised to see these K_MSEC() removed in this C file, were they being applied twice?

andyross replied:
More likely they were being used as terms in an expression (like "take half the blink period").  In general I tried to choose changes to minimize the size of the patch.  So if you had one timeout defined as a number but used unmodified in several of places, I'd put the macro in the declaration.  If it was a time value used with modification, I'd leave it as an integer in the header and change the usage to k_msleep(), etc...

andrewboie commented on include/sys_clock.h:
```diff
@@ -53,6 +53,8 @@ extern void z_enable_sys_clock(void);
 /* number of nanoseconds per second */
 #define NSEC_PER_SEC ((NSEC_PER_USEC) * (USEC_PER_MSEC) * (MSEC_PER_SEC))
 
+#define k_msleep(ms) k_sleep(ms)
```
I think it would be better to document this function and add it to our APIs alongside where k_sleep() is defined (which IIRC is in kernel.h). at minimum I think doxygen is needed.

andyross replied:
It gets that in an upcoming patch in the series, where it becomes a real function (because k_sleep() now takes a k_timeout_t).  I put the placeholders here just to grease the refactoring so everything bisects without weird splitting of features.

andrewboie commented on samples/basic/threads/src/main.c:
```diff
@@ -109,8 +109,8 @@ void uart_out(void)
 
 K_THREAD_DEFINE(blink1_id, STACKSIZE, blink1, NULL, NULL, NULL,
-		PRIORITY, 0, K_NO_WAIT);
+		PRIORITY, 0, 0);
```
Doesn't this eventually get propagated to k_thread_create()? Why does this have to be a 0 and not K_NO_WAIT?

andyross replied:
The timeout constructor macros are written in terms of the time unit conversion API, which are C functions.  C function calls, even pure ones that can be evaluated at compile time based on constant arguments, can't be used as a rvalue to a static initializer (C++ has constexpr for this, C never bothered).

We'd need a separate API to construct these, when it's really not that useful (the only use so far isn't to start a thread on a timer, it's to not start a thread at all).  The decision from the last time around was that we could address this with a new thread definition API (e.g. a "K_NO_START" flag).

andrewboie commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -5,6 +5,7 @@ config CMSIS_RTOS_V1
 	bool "CMSIS RTOS v1 API"
 	depends on THREAD_CUSTOM_DATA
 	depends on POLL
+	select LEGACY_TIMEOUT_API
```
we should probably file an enhancement GH ticket that cmsis v1/v2 and POSIX currently forces the legacy API

edit: net buffers and console as well. how much effort is it to convert these?

andyross replied:
Not a lot, really.  But it involves a little bit of design decision (i.e. does the subsystem want to expose k_timeout_t in its own API or stick with milliseconds and convert, etc..).  One of the sticking points last time was adding changes to subsystems without maintainer approval, so this one limits itself to subsystems that port with only trivial changes.

andrewboie replied:
OK. I think  we should have GH tickets just so we can track it.

andrewboie commented on tests/kernel/mem_protect/futex/prj.conf:
```diff
@@ -2,3 +2,4 @@ CONFIG_ZTEST=y
 CONFIG_IRQ_OFFLOAD=y
 CONFIG_TEST_USERSPACE=y
 CONFIG_MP_NUM_CPUS=1
+CONFIG_LEGACY_TIMEOUT_API=y
```
why does the futex test need this?

andyross replied:
It passes its own timeout value around via pointer arguments while assuming it's an integer, and would have been a minor port to the logic.

andrewboie commented on include/sys_clock.h:
```diff
@@ -53,6 +53,8 @@ extern void z_enable_sys_clock(void);
 #define NSEC_PER_SEC ((NSEC_PER_USEC) * (USEC_PER_MSEC) * (MSEC_PER_SEC))
 
+#define k_msleep(ms) k_sleep(ms)
+#define K_TIMEOUT_EQ(a, b) ((a) == (b))
```
this also needs doxygen

andyross commented:
Update to include k_uptime_ticks(), which was in the RFC but been forgotten.  Also sneak in the extra test code for absolute timeouts which got dropped in the first version.

nordic-krch commented on include/kernel.h:
```diff
@@ -773,7 +773,7 @@ extern void k_thread_foreach_unlocked(
  * @param prio Thread priority.
  * @param options Thread options.
- * @param delay Scheduling delay (in milliseconds), or K_NO_WAIT (for no delay).
+ * @param delay Scheduling delay, or K_NO_WAIT (for no delay).
```
im missing a place where timeout argument is well described. `k_timeout_t` in `sys_clock.h` is missing description at all and in the api it is very limited. I know that it is because of two modes but still maybe there should be something like: `see k_timeout_t for more details`.

andyross replied:
You're right, there should be a doc block on k_timeout_t.  Will add one.

nordic-krch commented on include/kernel.h:
```diff
@@ -879,6 +879,19 @@ __syscall int k_thread_stack_space_get(const struct k_thread *thread,
+ * This routine puts the current thread to sleep for @a duration,
+ * specified as a k_timeout_t object.
+ *
+ * @param ms Desired duration of sleep.
```
parameter should not be called `ms` anymore.

andyross replied:
Good point.

nordic-krch commented on include/kernel.h:
```diff
@@ -982,6 +998,43 @@ __syscall void k_thread_abort(k_tid_t thread);
+#ifdef CONFIG_SYS_CLOCK_EXISTS
+
+/**
+ * @brief Get time when a thread wakes up, in ticks
```
"in system ticks" would be better.

nordic-krch commented on include/kernel.h:
```diff
@@ -982,6 +998,43 @@ __syscall void k_thread_abort(k_tid_t thread);
+}
+
+/**
+ * @brief Get time remaining before a thread wakes up, in ticks
```
"in system ticks"

nordic-krch commented on include/kernel.h:
```diff
@@ -1532,7 +1625,79 @@ const char *k_thread_state_str(k_tid_t thread_id);
+#ifdef CONFIG_TIMEOUT_64BIT
+
+/**
+ * @brief Generates an absolute/uptime timeout value in ticks
```
shouldn't it be `from ticks` or `from absolute/uptime given in ticks`?

andyross replied:
Meh.  Grammatically: it's being documented as a constructor, not a conversion routine.  The timeout "is" the value in ticks.  But sure.

nordic-krch commented on include/kernel.h:
```diff
@@ -1532,7 +1625,79 @@ const char *k_thread_state_str(k_tid_t thread_id);
+#define K_TIMEOUT_ABS_TICKS(t) Z_TIMEOUT_TICKS(Z_TICK_ABS(MAX(t, 0)))
+
+/**
+ * @brief Generates an absolute/uptime timeout value in ms
```
shouldn't it be `from ms`  or `from absolute/uptime given in ms`?

nordic-krch commented on include/kernel.h:
```diff
@@ -1532,7 +1625,79 @@ const char *k_thread_state_str(k_tid_t thread_id);
+#define K_TIMEOUT_ABS_MS(t) K_TIMEOUT_ABS_TICKS(k_ms_to_ticks_ceil64(t))
+
+/**
+ * @brief Generates an absolute/uptime timeout value in us
```
same concern as above

nordic-krch commented on include/kernel.h:
```diff
@@ -1795,6 +1989,17 @@ static inline void *z_impl_k_timer_user_data_get(struct k_timer *timer)
+ *
+ * @return Current uptime in ticks.
+ */
+__syscall s64_t k_uptime_ticks(void);
```
why not `u64_t`?

andyross replied:
To conform with k_uptime_get() mostly, which was (probably -- it predates me, but I agree) done to prevent the application author from computing a time delta in unsigned units and tripping over the resulting overflow.

nordic-krch commented on include/kernel.h:
```diff
@@ -1810,7 +2015,10 @@ static inline void *z_impl_k_timer_user_data_get(struct k_timer *timer)
  * @return Current uptime in milliseconds.
  */
-__syscall s64_t k_uptime_get(void);
+static inline s64_t k_uptime_get(void)
```
in future that should be changed to u64_t as `k_ticks_to_ms_floor64` returns that anyway.

andyross replied:
It really should not, but we can have that fight later :)

nordic-krch commented on include/sys_clock.h:
```diff
@@ -28,6 +28,82 @@
+ * @addtogroup clock_apis
+ * @{
+ */
+
```
i'm missing here a common description of `k_timeout_t`. What it represents, what are the special values, etc. Recommendations (like using K_MSEC). This could be referenced by other API's

nordic-krch commented on include/sys_clock.h:
```diff
@@ -28,6 +28,82 @@
+typedef s32_t k_timeout_t;
+#define K_TIMEOUT_EQ(a, b) ((a) == (b))
+#define Z_TIMEOUT_NO_WAIT 0
+#define Z_TIMEOUT_TICKS(t) k_ticks_to_ms_ceil32(t)
```
should that be just `(k_timeout_t)t`?

andyross replied:
No, in legacy mode the value stored in the k_timeout_t is a count of milliseconds.

nordic-krch replied:
yes, but this macro is used in the context where `t` is ticks value:
https://github.com/andyross/zephyr/blob/10aaf24035586b315645d3cd9146ef52362cec68/kernel/sched.c#L1162
https://github.com/andyross/zephyr/blob/10aaf24035586b315645d3cd9146ef52362cec68/samples/scheduler/metairq_dispatch/src/msgdev.c#L78

andyross replied:
That's correct.  The final argument to z_add_timeout() has changed: in legacy code it's a 32 bit count of ticks (not milliseconds) which ipc mechanisms compute from application-provided arguments.  In new code it's a k_timeout_t that is constructed in the application code.

nordic-krch commented on include/sys_clock.h:
```diff
@@ -28,6 +28,82 @@
+ * t for all inputs, and that the representation of K_TICKS_FOREVER is
+ * the same value in both spaces!  Clever, huh?
+ */
+#define Z_TICK_ABS(t) (K_TICKS_FOREVER - 1 - (t))
```
maybe something with `GET_ABS` in name. Otherwise it can be confused with converting t to absolute (positive) value.

andyross replied:
It's a little subtle in general, honestly, thus that comment.  I don't know that a simple name is going to really help that much -- this is an internal gadget used in like three places.

nordic-krch commented on include/sys_clock.h:
```diff
@@ -130,6 +206,8 @@ s64_t z_tick_get(void);
 #define z_tick_get_32() (0)
 #endif
 
+u64_t z_timeout_end_calc(k_timeout_t timeout);
```
what does this function do? Missing dox.

andyross replied:
There's a big comment above it's implementation describing what it does, with a warning about how to use it correctly.  In general we don't doxygen internal functions.  It's declared here only so it can be used in an inline.

nordic-krch replied:
ok, i found the comment in the implementation. Imo, such comments should be in the header even though these functions are not public.

andyross replied:
Yeah, I think we have to agree to disagree on this one.  I really don't think putting documentation for internal functions into public headers is a good idea.

nordic-krch commented on kernel/queue.c:
```diff
@@ -293,45 +293,32 @@ int k_queue_merge_slist(struct k_queue *queue, sys_slist_t *list)
-			return NULL;
-		}
+	event.state = K_POLL_STATE_NOT_READY;
+	err = k_poll(&event, 1, timeout);
```
can you explain why loop is no longer needed?

andyross replied:
It was never needed.  The assumption seemed to be that k_poll() might fail spuriously and be retried, but k_poll() doesn't fail in a way that requires retries.  We should be propagating any internal errors, not hiding them.

And in this case fixing the design bug was easier than trying to refactor it to do retries over a timeout.

nordic-krch commented on kernel/sched.c:
```diff
@@ -562,20 +562,16 @@ static void pend(struct k_thread *thread, _wait_q_t *wait_q, s32_t timeout)
+#endif
 
-		z_add_thread_timeout(thread, ticks);
+		z_add_thread_timeout(thread, timeout);
```
there is some inconsistency there: `z_add_thread_timeout` takes `k_timeout_t` as the second argument. In legacy mode that means milliseconds and yet it seems that it takes ticks since ms to ticks conversion occurs.

Why not do that conversion for legacy mode in `z_add_timeout`?

andyross replied:
Mostly because I'm trying not to change the legacy mode code paths too much -- the code used to be done in the various subsystems above this function, so I left it that way.  The new code does, indeed, fold all the conversion code to a single place (the timeout constructor macros) and all the tick alignment to z_add_timeout().

nordic-krch replied:
legacy will stay for long maybe it would be good to have it clean. Looking at places where z_add_timeout is used, it seems to me that it could take k_timeout_t for legacy code as well. There is only one place where `z_add_timeout` is called with ticks without earlier conversion from milliseconds: `msgdev.c` and it is a sample using internal api so i guess it could be solved somehow (e.g. force new timeout scheme only).

andyross replied:
I'd really prefer to do that cleanup later.  I really don't think keeping a few 2-5 line blocks of code unchanged under an #if is that much of a burden.  And the intent is that LEGACY goes away in a release or two anyway.

nordic-krch replied:
can you please add comment in `z_add_timeout` explaining why for legacy implementation value stored in `k_timeout_t` is not actual k_timeout_t but ticks. Someone else looking at that code may get confused (someone else can be me or you in few months :)).

nordic-krch commented on kernel/timeout.c:
```diff
@@ -83,8 +83,18 @@ static s32_t next_timeout(void)
 }
 
-void z_add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)
+void z_add_timeout(struct _timeout *to, _timeout_func_t fn, k_timeout_t t)
```
could it take `k_ticks_t` as argument?

andyross replied:
Yes, but why?  In the new design, it just passes through a k_timeout_t that was constructed all the way up in application code without having to convert or inspect anything in the intervening layers.  That's better, I'd argue.

nordic-krch commented on kernel/timeout.c:
```diff
@@ -83,8 +83,18 @@ static s32_t next_timeout(void)
 }
 
-void z_add_timeout(struct _timeout *to, _timeout_func_t fn, s32_t ticks)
+void z_add_timeout(struct _timeout *to, _timeout_func_t fn, k_timeout_t t)
```
later on local variable `struct _timeout *t;` is created. Those names should not conflict.

andyross replied:
Good point.  I'm wondering why that collision isn't a warning....  Will fix.

nordic-krch commented on kernel/timeout.c:
```diff
@@ -128,29 +138,50 @@ int z_abort_timeout(struct _timeout *to)
-			if (timeout == t) {
-				break;
-			}
+	for (struct _timeout *t = first(); t != NULL; t = next(t)) {
```
any reason for removing lock? does not look like related to PR main goal.

andyross replied:
Whoops, this was a rebase error.  Will fix.

andyross replied:
Actually no, this is fine.  The lock got moved up a level so the caller could add its own math inside the lock.

andyross commented on kernel/Kconfig:
```diff
@@ -570,6 +570,26 @@ config SYS_CLOCK_EXISTS
+
+config TIMEOUT_64BIT
+	bool
+	default y if !LEGACY_TIMEOUT_API
```
Fair enough.  Though it needs both, since depends won't set it to y.

nordic-krch replied:
yes, it should be `depends on ... + default y`

andyross commented on drivers/i2s/Kconfig:
```diff
@@ -8,6 +8,7 @@
 #
 menuconfig I2S
 	bool "I2S bus drivers"
+	select LEGACY_TIMEOUT_API
```
There are some docs in the commit message.  Broadly subsystems that expose their own timeouts via APIs (e.g. they have functions that can accept K_FOREVER, etc...) need to decide how to do that -- they can either keep their current APIs as millisecond counts, or expose k_timeout_t directly (the latter requires a bit more work).  Andrew asked for GH issues to be filed for each of these, which I'll do.

nordic-krch replied:
Fair enough, i guess that they will be attacked one by one to enable new timeout API.

andyross commented on kernel/thread.c:
```diff
@@ -386,15 +386,17 @@ static inline void z_vrfy_k_thread_start(struct k_thread *thread)
 	} else {
-		s32_t ticks = _TICK_ALIGN + k_ms_to_ticks_ceil32(delay);
+#ifdef CONFIG_LEGACY_TIMEOUT_API
+		delay = _TICK_ALIGN + k_ms_to_ticks_ceil32(delay);
```
It now happens  at a single place in z_add_timeout().

andyross commented on tests/boards/intel_s1000_crb/src/dma_test.c:
```diff
@@ -42,10 +42,10 @@
 
 /* delay between greetings (in ms) */
-#define SLEEPTIME               500
+#define SLEEPTIME               K_MSEC(500)
```
Both are valid APIs.  You can use k_sleep() with a proper timeout value, or call k_msleep()/k_usleep() with an integer value you calculate yourself.    Generally I tried to refactor legacy code in a way that naturally matches what they were doing (which is to say: I mostly tried to optimize the line count of the patch).

andyross commented on tests/kernel/mem_protect/futex/src/main.c:
```diff
@@ -55,7 +55,7 @@ void futex_wait_task(void *p1, void *p2, void *p3)
 	int time_val = *(int *)p1;
 
-	zassert_true(time_val >= K_FOREVER, "invalid timeout parameter");
+	zassert_true(time_val >= (int)K_FOREVER, "invalid timeout parameter");
```
Indeed, will fix.

andyross commented on samples/scheduler/metairq_dispatch/src/msgdev.c:
```diff
@@ -77,7 +77,11 @@ static void timeout_reset(void)
 	u32_t ticks = rand32() % MAX_EVENT_DELAY_TICKS;
 
+#ifdef CONFIG_LEGACY_TIMEOUT_API
 	z_add_timeout(&timeout, dev_timer_expired, ticks);
```
Not understanding.  Yes, of course, but nothing in the legacy API uses that macro, it's new.  So why bother supporting it in legacy mode?

nordic-krch replied:
what i meant was that if `Z_TIMEOUT_TICKS` would be created for legacy then ifdef would not be needed. 

Currently, it is defined as `#define Z_TIMEOUT_TICKS(t) k_ticks_to_ms_ceil32(t)`. What if it is defined as `#define Z_TIMEOUT_TICKS(t) t`. That would simplify also use in `shed.c:1161` to something like: `timeout = Z_TIMEOUT_TICKS(ticks + (IF_ENABLED(LEGACY) ? _TICK_ALIGN : 0));`

andyross commented on kernel/timer.c:
```diff
@@ -105,29 +107,45 @@ void k_timer_init(struct k_timer *timer,
+	 * timer_api test relies on this behavior.
+	 */
+	period.ticks = MAX(period.ticks - 1, 0);
+	if (Z_TICK_ABS(duration.ticks) < 0) {
```
There is some documentation in the Z_TICK_ABS() macro explaining the convention.  I can add the convenience wrapper if you like.

nordic-krch commented on kernel/timeout.c:
```diff
@@ -83,8 +83,18 @@ static s32_t next_timeout(void)
+	k_ticks_t ticks = t.ticks + 1;
+
+	if (IS_ENABLED(CONFIG_TIMEOUT_64BIT) && Z_TICK_ABS(ticks) >= 0) {
+		ticks = Z_TICK_ABS(ticks) - curr_tick;
```
seems to me that it should be `- (curr_tick + elapsed())`

andyross replied:
Arrgh, right as I was about to push.  Yes, this is a bug.  Will fix and update.

Commit:
kernel/timeout: Doc cleanup

Add a doc string for k_timeout_t.  Rephrase a few other items for
clarity.  Fix some syntax goofs.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebased again around a minor collision.  Cleaned up the long line warnings from checkpatch.  Added the discussed doc changes and a few other minor things from review.  Fixed the elapsed() bug @nordic-krch  found.

Should be good to go.

nordic-krch commented:
I think it would be good to align timeout parameter documentation in `kernel.h`

For example `k_thread_create` delay is described as `Scheduling delay, or K_NO_WAIT (for no delay)`.
but I noticed that here https://github.com/zephyrproject-rtos/zephyr/blob/698000f40c6e9f44fbc7753b259ae5e92ba90fc2/tests/kernel/threads/thread_apis/src/test_threads_spawn.c#L119 `k_thread_create` is called with `K_FOREVER`.

Some examples:
- `Non-negative waiting period to obtain a data item or one of the special values K_NO_WAIT and                K_FOREVER.`
- `Delay before submitting the work item` (no mention about special values)

Imo, they could all refer to `k_timeout_t`. Like:
- `Waiting period to obtain  a data item. See k_timeout_t for details.`
- `Delay before submitting the work item. See k_timeout_t for details.`
- etc.

nordic-krch commented:
@andyross what about CI? Any chance to get results there?

andyross commented:
Are those really something that should block merge?  The documentation as written is correct.  I agree it could be improved.  Can't we file an issue on that or just submit it personally post-merge?

As far as CI: no, PRs that touch this many tests don't complete.  Someone has to run it manually before merge.

andyross commented:
Rebased

nashif commented:
> @andyross what about CI? Any chance to get results there?

@nordic-krch given the amount of the changes CI times out here, this will go through a manual run before merge. If you see any build/test issues related to this change, please raise them. Approval of this PR should not depend on CI passing.

The only alterntive is to raise the number of shippable minions  in .shippable.yml and let CI complete and then remove the CI change before the merge. @andyross You can do that if you feel like it.

nashif commented:
just noticed this one:

```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/xlnx_psttc_timer.c:164:12: error: invalid operands to binary == (have 's32_t' {aka 'int'} and 'k_timeout_t' {aka 'struct <anonymous>'})
  164 |  if (ticks == K_FOREVER) {
      |            ^~
```

andyross commented:
Correct some rebase crosses in built-but-not-run platform and one spot in tests/kernel/context that wasn't compatible and literally just merged.

nordic-krch commented:
there are 2 issues remaining:
- `loapic_timer.c`
- mcuboot. I've created PR for mcuboot: https://github.com/JuulLabs-OSS/mcuboot/pull/693
However, mcuboot part has some dependencies since #23625 required mcuboot changes which already got merged (https://github.com/JuulLabs-OSS/mcuboot/pull/682)

andyross commented:
Fixed up the acrn failure (shouldn't acrn be built under sanitycheck if it's run under CI?  What broke there?)

Similarly: when did mcuboot move out of tree?  I don't even know how to run those tests, why are they in CI?

nashif commented:
>     * mcuboot. I've created PR for mcuboot: [JuulLabs-OSS/mcuboot#693](https://github.com/JuulLabs-OSS/mcuboot/pull/693)

I do not think it is a good idea to be depending on upstream PRs, I would first submit this and make it work in the repo maintained in west and do the upstreaming in parallel, otherwise PRS will be stuck for a long time depending on how fast upstream reacts.

andyross commented:
Rebase to resolve collisions with commit a4c919064973 and commit 84921c53b1b5 which snuck around this one into master.

Note I did run this through a full sanitycheck myself, just spot checked the relevant tests (kernel/threads and kernel/interrupt) kernel/common across the standard platforms.

carlescufi commented:
> > ```
> > * mcuboot. I've created PR for mcuboot: [JuulLabs-OSS/mcuboot#693](https://github.com/JuulLabs-OSS/mcuboot/pull/693)
> > ```
> 
> I do not think it is a good idea to be depending on upstream PRs, I would first submit this and make it work in the repo maintained in west and do the upstreaming in parallel, otherwise PRS will be stuck for a long time depending on how fast upstream reacts.

What we've been doing for tightly-coupled projects like MCUboot is to send a PR upstream, point to that PR and remote in the `west.yml` in the zephyr PR, then merge the upstream MCUboot PR, then updated our MCUboot copy with the latest MCUboot, then point the `west.yml` in this PR to our MCUboot.
@nvlsianpu can help with that, or @nordic-krch or myself can do that

andyross commented:
Er, excuse me: I did **not** run this through sanitycheck before pushing.

As far as that mcuboot PR, it shows merged now.  I'm hoping we're OK now.

nashif commented:
DNM so we remember to remove the shippable hack.

andyross commented:
One more to clean up another crossed issue with the new cavs_timer driver.  This also includes a build fix to an upstream bug with cc26x2r1_launchxl  that is merely unmasked here.  It's not a timer driver thing at all, but it's a simple and obvious patch.

Note that we're still seeing a failure in mcuboot though.  @nordic-krch , can you take a look at that?

andyross commented:
OK, we're good now.  Just the one mcuboot failure left.

vanti commented:
Fyi the cc26x2r1_launchxl issue is being addressed in #23929.

carlescufi commented:
@andyross please rebase, the mcuboot build should be fixed now

nordic-krch commented:
Remaining issue with mcuboot in PR: https://github.com/JuulLabs-OSS/mcuboot/pull/700

andyross commented:
Rebased (actually the only collision was that last patch to cc26x2r1_launchxl , which made mine unnecessary so I just removed it).
